File: Sources/PalmierPro/Editor/RippleEngine.swift
import Foundation

/// A proposed new start frame for a single clip, produced by the ripple engine
/// and applied by the caller.
struct ClipShift: Equatable, Sendable {
    let clipId: String
    let newStartFrame: Int
}

/// A half-open `[start, end)` frame interval on a single track. Used to describe
/// the gaps that a ripple edit needs to close.
struct FrameRange: Equatable, Sendable {
    let start: Int
    let end: Int
    var length: Int { end - start }
}

/// A user-selected empty gap on a single track
struct GapSelection: Equatable, Sendable {
    let trackIndex: Int
    let range: FrameRange
}

/// Pure functions for ripple editing: computing how clips shift after
/// insertions or deletions.
enum RippleEngine {

    /// After removing clips from a track, compute new start frames for
    /// remaining clips that should shift backward to close the gap.
    static func computeRippleShifts(clips: [Clip], removedIds: Set<String>) -> [ClipShift] {
        let removedRanges = clips
            .filter { removedIds.contains($0.id) }
            .map { FrameRange(start: $0.startFrame, end: $0.endFrame) }
        return computeRippleShiftsForRanges(
            clips: clips.filter { !removedIds.contains($0.id) },
            removedRanges: removedRanges
        )
    }

    /// Shift clips leftward to close the gaps defined by `removedRanges`.
    /// Used when ranges come from a different track (sync-locked ripple).
    static func computeRippleShiftsForRanges(clips: [Clip], removedRanges: [FrameRange]) -> [ClipShift] {
        let merged = mergeRanges(removedRanges)
        guard !merged.isEmpty else { return [] }

        var shifts: [ClipShift] = []
        for clip in clips.sorted(by: { $0.startFrame < $1.startFrame }) {
            let shift = merged
                .filter { $0.end <= clip.startFrame }
                .reduce(0) { $0 + $1.length }
            if shift > 0 {
                shifts.append(ClipShift(clipId: clip.id, newStartFrame: clip.startFrame - shift))
            }
        }
        return shifts
    }

    /// Push all clips at or after `insertFrame` forward by `pushAmount` frames.
    static func computeRipplePush(
        clips: [Clip],
        insertFrame: Int,
        pushAmount: Int,
        excludeIds: Set<String> = []
    ) -> [ClipShift] {
        clips
            .filter { !excludeIds.contains($0.id) && $0.startFrame >= insertFrame }
            .map { ClipShift(clipId: $0.id, newStartFrame: $0.startFrame + pushAmount) }
    }

    // MARK: - Helpers

    static func mergeRanges(_ ranges: [FrameRange]) -> [FrameRange] {
        let sorted = ranges.sorted { $0.start < $1.start }
        var merged: [FrameRange] = []
        for range in sorted {
            if let last = merged.last, range.start <= last.end {
                merged[merged.count - 1] = FrameRange(start: last.start, end: max(last.end, range.end))
            } else {
                merged.append(range)
            }
        }
        return merged
    }
}

File: windows/src/PalmierPro.Core/Editing/RippleEngine.cs
using PalmierPro.Core.Models;

namespace PalmierPro.Core.Editing;

/// <summary>Half-open frame range [Start, End).</summary>
public readonly record struct FrameRange(int Start, int End)
{
    public int Length => End - Start;
    public bool IsValid => End > Start;
    public bool Contains(int frame) => frame >= Start && frame < End;
}

public readonly record struct ClipShift(string ClipId, int NewStartFrame);

/// <summary>
/// Pure ripple math mirroring the Mac RippleEngine: closing removed ranges shifts
/// every clip left by the total length of merged ranges entirely to its left, and
/// insertion pushes clips at or after the insert frame right.
/// </summary>
public static class RippleEngine
{
    public static List<FrameRange> MergeRanges(IReadOnlyList<FrameRange> ranges)
    {
        var valid = ranges.Where(r => r.IsValid).OrderBy(r => r.Start).ToList();
        var merged = new List<FrameRange>();
        foreach (var range in valid)
        {
            if (merged.Count > 0 && range.Start <= merged[^1].End)
            {
                merged[^1] = new FrameRange(merged[^1].Start, Math.Max(merged[^1].End, range.End));
            }
            else
            {
                merged.Add(range);
            }
        }
        return merged;
    }

    public static List<ClipShift> ComputeRippleShiftsForRanges(
        IReadOnlyList<Clip> clips, IReadOnlyList<FrameRange> removedRanges)
    {
        var merged = MergeRanges(removedRanges);
        var shifts = new List<ClipShift>();
        foreach (var clip in clips.OrderBy(c => c.StartFrame))
        {
            var shift = merged.Where(r => r.End <= clip.StartFrame).Sum(r => r.Length);
            if (shift > 0) shifts.Add(new ClipShift(clip.Id, clip.StartFrame - shift));
        }
        return shifts;
    }

    public static List<ClipShift> ComputePushShifts(
        IReadOnlyList<Clip> clips, int insertFrame, int pushAmount)
    {
        if (pushAmount <= 0) return [];
        return clips
            .Where(c => c.StartFrame >= insertFrame)
            .Select(c => new ClipShift(c.Id, c.StartFrame + pushAmount))
            .ToList();
    }
}
